File: Bamilo/TrackerManager.m
////
////  TrackerManager.m
////  Bamilo
////
////  Created by Narbeh Mirzaei on 4/8/17.
////  Copyright © 2017 Rocket Internet. All rights reserved.
////
//
//#import "TrackerManager.h"
//
//@implementation TrackerManager
//
//static NSMutableArray <BaseTracker *> *_trackers;
//
//#pragma mark - Public Methods
//+ (NSMutableArray *)addTracker:(BaseTracker *)tracker {
//    if(_trackers == nil) {
//        _trackers = [NSMutableArray<BaseTracker *> new];
//    }
//    [_trackers addObject:tracker];
//    return _trackers;
//}
//
//
//+ (void)postEventViaSelector:(SEL)selector attributes:(NSDictionary *)attributes {
//    
//}
////+ (void)postEvent:(NSDictionary *)attributes forName:(NSString *)name {
////    for (id tracker in _trackers) {
////        if([tracker conformsToProtocol:@protocol(EventTrackerProtocol)] && [tracker isEventEligable:name]) {
////            [tracker postEvent:attributes forName:name];
////        }
////    }
////}
//
//+ (void)trackScreenName:(NSString *)screenName {
//    for (id tracker in _trackers) {
//        if([tracker conformsToProtocol:@protocol(ScreenTrackerProtocol)]) {
//            [tracker trackScreenName:screenName];
//        }
//    }
//}
//
//+ (void)sendTags:(NSDictionary *)tags completion:(TagTrackerCompletion)completion {
//    for(id<TagTrackerProtocol> tracker in _trackers) {
//        if([tracker conformsToProtocol:@protocol(TagTrackerProtocol)]) {
//            [tracker sendTags:tags completion:completion];   
//        }
//    }
//}
//
//@end

File: TrackerManager.swift
//
//  TrackerManager.swift
//  Bamilo
//
//  Created by Ali Saeedifar on 7/5/17.
//  Copyright © 2017 Rocket Internet. All rights reserved.
//

@objcMembers class TrackerManager: NSObject {
    
    static var trackers = [BaseTracker]()
    
    class func addTracker(tracker: BaseTracker) {
        self.trackers.append(tracker)
    }
    
    class func postEvent(selector: Selector, attributes: EventAttributeType) {
        trackers.forEach { (tracker) in
            if tracker.conforms(to: EventTrackerProtocol.self) && tracker.responds(to: selector){
                tracker.perform(selector, with: attributes)
            }
        }
    }
    
    class func trackLoadTime(screenName: String, interval: NSNumber, label: String) {
        trackers.forEach { (tracker) in
            (tracker as? ScreenTrackerProtocol)?.trackLoadTime?(screenName: screenName, interval: interval, label: label)
        }
    }
    
    class func trackScreenName(screenName:String) {
        self.trackers.forEach { (tracker) in
            (tracker as? ScreenTrackerProtocol)?.trackScreenName(screenName: screenName)
        }
    }
    
    class func sendTag(tags: [String:Any], completion: @escaping TagTrackerCompletion) {
        self.trackers.forEach { (tracker) in
            (tracker as? TagTrackerProtocol)?.sendTags(tags, completion: completion)
        }
    }
}
